src/dmd/mtype.d

```diff
@@ -386,20 +386,27 @@ extern (C++) abstract class Type : ASTNode
     MOD mod; // modifiers MODxxxx
     char* deco;
 
-    /* These are cached values that are lazily evaluated by constOf(), immutableOf(), etc.
-     * They should not be referenced by anybody but mtype.c.
-     * They can be NULL if not lazily evaluated yet.
-     * Note that there is no "shared immutable", because that is just immutable
-     * Naked == no MOD bits
-     */
-    Type cto;       // MODFlags.const_                 ? naked version of this type : const version
-    Type ito;       // MODFlags.immutable_             ? naked version of this type : immutable version
-    Type sto;       // MODFlags.shared_                ? naked version of this type : shared mutable version
-    Type scto;      // MODFlags.shared_ | MODFlags.const_     ? naked version of this type : shared const version
-    Type wto;       // MODFlags.wild                  ? naked version of this type : wild version
-    Type wcto;      // MODFlags.wildconst             ? naked version of this type : wild const version
-    Type swto;      // MODFlags.shared_ | MODFlags.wild      ? naked version of this type : shared wild version
-    Type swcto;     // MODFlags.shared_ | MODFlags.wildconst ? naked version of this type : shared wild const version
+    static struct Mcache
```

Comment:
No chance to factor this together with the other? Looks like some pretty nasty duplication...

Reply:
That whole astbase.d is duplication.

Reply:
It would be nice if astbase were to go away.  But alas, there probably is someone out there who found a use for it outside the dmd testsuite.

Reply:
Deprecate it.

Reply:
The idea was to remove the rest of the "real" AST and leave only AST base. I.e. move all of the logic out of the AST classes, which should remain as pure data containers.

Reply:
ASTbase lumps together everything in one big mess, it would be horrid if the front-end were to start actually being dependent on it.